topics/dapp-wallet-msg.md

```diff
@@ -0,0 +1,110 @@
+    - a list of networks on which the relying party plans to operate (`networks`),
+    - a list of permission scopes it requires (`scopes`).
+2. The wallet receives the message, presents the details of the to-be-established connection to the user and asks the user to select the account that will be paired in response.
+    - If the user approves the request, the wallet sends `permission_response` back to the relying party. The message includes:
+        - basic information about the wallet (`appMetadata`),
+        - a list of networks on which the user has agreed the relying part can operate, should be a subset of the equivalent from `permission_request` (`networks`),
+        - a list of permission scopes the user has agreed the relying party can be granted, should be a subset of the equivalent from `permission_request` (`scopes`)
+        - the account details (princial or public key (*TBD what is required*) and optional subaccount) of the user selected account (`account`)
+    - If the user rejects the request, the wallet sends `error` back to the relying party with appropriate explanation.
+
+```
+{ // Request
+    type: "permission_request"
+    appMetadata: /* dApp */ {
+        name: String
+        icon: Optional<String>
+    }
+    networks: Array<{
+        type: "mainnet" | "testnet"
+        name: Optional<String>
+        rpcUrl: Optional<String>
+    }>
+    scopes: Array<"canister_call">
+}
+
+{ // Response
+    type: "permission_response"
+    appMetadata: /* wallet */ {
+        name: String
+        icon: Optional<String>
+    }
+    networks: Array<{
+        type: "mainnet" | "testnet"
+        name: Optional<String>
+        rpcUrl: Optional<String>
+    }>
+    scopes: Array<"canister_call">
+    account: {
+        owner: String /* principal / publicKey */
+        subaccount: Optional<String>
+    }
+}
+```
+
+## Canister Call
+
+Once the connection between the relying party and wallet is established, the relying party can request the wallet to execute canister calls:
+1. The relying party sends `canister_call_request`. The message incudes:
+    - the network details on which the call should be executed,
+    - the ID of the target canister,
+    - the name of the call method to be executed,
+    - the arguments for the call.
+2. The wallet receives the request, validates if the relying party has permission to request the action and processes the message (in accordance with the ICRC-21 specification).
+    - If the user approves the request, the wallet sends the call to the IC and, upon receiving a successful response, sends `canister_call_response` back to the relying party. The message includes:
+        - the response from IC,
+        - optionally (*TBD*), the parameters (or their subset) of the request.
+    - If the user rejects the request or wallet fails to complete the requested action for any reason, the wallet sends `error` back to the relying party with appropriate explanation.
+
+```
+{ // Request
```

Comment:
> I've just noticed that the ICRC-21 draft says that the execution of the call is still done by the wallet and only the result of the call is handed back to the relying party, while the RFP-7 document specifies that the wallet is supposed to return only signatures for the call itself plus read_state that can come after.

Sorry for the discrepancy between RFP-7 and ICRC-21. I made some last minute changes to ICRC-21 after RFP-7 went out. We should go with the ICRC-21 variant (the wallet submits the call to the IC).
The changes to ICRC-21 were to address feedback from the internal product security team, that the wallet should be able to observe the interaction with IC in order to prevent malicious applications being able to lie to users about the outcome of submitting the signed call to the IC (i.e. one could imagine a malicious marketplace faking network loss in order to trick the user to sign a second payment transaction).

But in that model, for the relying party to verify that the wallet did not modify the request before submitting it to the IC (and to verify that the response is genuine), the relying party needs the full [certificate](https://internetcomputer.org/docs/current/references/ic-interface-spec/#certificate) as returned by the `read_state` call and the full content map in order to calculate the [request id](https://internetcomputer.org/docs/current/references/ic-interface-spec/#request-id).

I have no strong opinion, whether the relying submits the full content map as part of the request or the wallet returns the matching content map back to the relying party together with the certificate, I think both could work fine.

> Regarding the types, I agree that they should be the same as the specification (arg and response should be bytes representing encoded values), but we have to keep in mind that the messages have to be serializable, sendable via different kind of transports, and readable on various platforms, therefore instead of using an actual byte array, I would use a hex string representation.

Good point! Is there an argument to be made against base64? This would make the strings much more compact.

Reply:
> We should go with the ICRC-21 variant (the wallet submits the call to the IC).

👍 

> But in that model, for the relying party to verify that the wallet did not modify the request before submitting it to the IC (and to verify that the response is genuine), the relying party needs the full [certificate](https://internetcomputer.org/docs/current/references/ic-interface-spec/#certificate) as returned by the read_state call and the full content map in order to calculate the [request id](https://internetcomputer.org/docs/current/references/ic-interface-spec/#request-id).

Just for me to understand, why should the relying party not trust the response from the wallet and need to verify that the request wasn't modified? 🤔 If the wallet was malicious and meant harm, the damage would be already done as the request would have been submitted before the relying party could verify anything. 

What's more, I always believed that *the relying party* is what should never be trusted, therefore the wallet (or rather the user through the wallet) should always have the last word in the requests, being even allowed to adjust them to the user's liking. A good example could be token transfers, where you can specify the fee you're willing to pay for the transaction. It wouldn't necessarily require a malicious act to send over a request with the fee set already and set to a value that the user finds inadequate and wants to change before submitting.

I can see that maybe from the UX perspective it could be useful to check if the response is what was expected to better communicate where the problem originated from, if there was any. But if the response was truly incorrect, it would get surfaced sooner or later while decoding, I think.

I mean, I'm not exactly advocating against such a model, but as a wallet developer, I find it quite complex and cumbersome (especially with the lack of support in the Dfinity libraries, but this is, I guess, something that could be fixed in the future releases) and I want to be sure I fully understand the reason behind such a requirement 😅 

> Is there an argument to be made against base64? This would make the strings much more compact.

Oh, you're right! Base64 should do a better job 😅

Reply:
It's about the relying party not trusting a specific wallet and the user. 

Imagine a gated application that only allows access if the user owns an asset with some specific canister. And furthermore, that canister does not offer a way to publicly query who owns assets.

In that case, the relying party would only need to trust that other canister and _not_ the wallet, if the response from the canister is handed through in such a way that the relying party can verify its authenticity.

In general, the wallet is under the users control and _should not_ be trusted anyway. I.e. if users can gain anything from using a wallet that does not forward genuine responses, they will absolutely use that.

And if the responses from the wallet are not verifiable, that just means that all interactions with the wallet need to be double checked by making out of band queries to the IC. Which also restricts the use cases to only those that actually can be checked publicly (since the relying party must not use the wallet for the check).

> What's more, I always believed that the relying party is what should never be trusted, therefore the wallet (or rather the user through the wallet) should always have the last word in the requests, being even allowed to adjust them to the user's liking.

I'm not advocating against allowing the user to tinker with the request within the wallet. This is useful, agreed. But the protocol should not support lying. 😉 

My point is that the relying party should receive all the information necessary to _verify_ the interaction with the IC. Having the wallet do A, but the tell the relying party it did B should lead to a protocol error (because the verification on the relying party side will fail).

Reply:
Ok, I think I understand now and I'm convinced 😄 I still keep forgetting that IC on many levels is different than most of the other networks and, for example, a lot of user data is not necessarily public.

> I have no strong opinion, whether the relying submits the full content map as part of the request or the wallet returns the matching content map back to the relying party together with the certificate, I think both could work fine.

I'd make that the wallet's responsibility to include the content map and certificate in the response, then. I think this should give more flexibility (if we can even speak of any 🤔) as the relying party could ask for a canister call result in the simplest way possible, let the wallet do its job, and only then worry about how the request was made and if the relying party is fine with it:

```
{ // Request
    type: "canister_call_request"
    network: {
        chainId: String
        name: Optional<String>
        rpcUrl: Optional<String>
    }
    canisterId: String
    method: String
    args: String /* encoded, base64 */
}

{ // Response
    type: "canister_call_response"
    network: {
         type: "mainnet" | "testnet"
         name: Optional<String>
         rpcUrl: Optional<String>
    }
    response: String /* encoded, base64 */
    contentMap: {
        request_type: String
        sender: String /* principal */
        nonce: Optional<String /* base64 */>
        ingress_expiry: Int
        canister_id: String /* base64 */
        method_name: String
        arg: String /* base64 */
    }
    certificate: {
        tree: // TODO: type
        signature: String /* base64 */
        delegation: // TODO: type
    }
}
```

Reply:
ah, another thing: The certificate already includes the response. So this response type duplicates that information. As such, I would only include the certificate (and remove the extra `response` field).

Reply:
> the network block on the response is still the old type.

Ah yes, thanks for pointing that out 😅 I corrected it.

Ok, I make sure to update the document before the next meeting.

---

> The certificate already includes the response. So this response type duplicates that information. As such, I would only include the certificate (and remove the extra response field).

Does it? Is it included in the tree?

Reply:
> Does it? Is it included in the tree?

Yes, see [here](https://internetcomputer.org/docs/current/references/ic-interface-spec/#state-tree-request-status). The path to the response `blob` is `/request_status/<request_id>/reply`.

Reply:
I see! I wasn't aware 😅 I'll remove the `response` field right away